ssz_serialization/proofs.nim

```diff
@@ -252,6 +257,106 @@ func verify_merkle_multiproof*(
   if calc.isErr: return false
   calc.get == root
 
+func verify_merkle_multiproof*(
+    leaves: openArray[Digest],
+    proof: openArray[Digest],
+    indices: static openArray[GeneralizedIndex],
+    root: Digest): bool =
+  const helper_indices = get_helper_indices(indices)
+  let calc = calculate_multi_merkle_root(leaves, proof, indices, helper_indices)
+  if calc.isErr: return false
+  calc.get == root
+
+# https://github.com/ethereum/consensus-specs/blob/v1.1.6/tests/core/pyspec/eth2spec/test/helpers/merkle.py#L4-L21
+func build_proof*(
+    anchor: auto,
+    indices: openArray[GeneralizedIndex],
+    helper_indices: openArray[GeneralizedIndex],
+    proof: var openArray[Digest]): Result[void, string] =
+  doAssert proof.len == helper_indices.len
+  ? check_multiproof_acceptable(indices)
+  hash_tree_root(anchor, helper_indices, proof)
```

Comment:
I find the name `hash_tree_root` somewhat confusing for this use case. Why don't we use a more explicit name that references the generation of proofs.

Reply:
A merkle proof is just a combination of roots, one for each `helper_index`. The `helper_indices` are computed with `get_helper_indices` from the requested `indices`, and then `hash_tree_root` is called on each of those `helper_indices` to compute the roots for each of them. To avoid O(n^2), all indices are passed to `hash_tree_root` in one go. However, `hash_tree_root` does not know how the roots will be used, I think the name is alright as it just gains an overload to compute multiple roots in one go, and being able to request non-1 generalized indices.

The function here is called `build_proof`, it's the same name that the EF consensus spec use for building proofs in their test suite.

We have `build_proof` overloads for dynamic data (e.g., client requested proofs), and also for static indices where helper indices can be pre-computed at compile time (e.g., light client update production). We also have `build_proof` overloads for cases where helper indices are already known, to avoid repeat computations.

Reply:
I see, the explanation makes sense. Thanks!